rcl_action/src/rcl_action/action_server.c

```diff
@@ -554,20 +640,23 @@ rcl_action_expire_goals(
       continue;
     }
     goal_time = _goal_info_stamp_to_nanosec(info_ptr);
-    assert(current_time > goal_time);
     if ((current_time - goal_time) > timeout) {
       // Stop tracking goal handle
       // Fill in any gaps left in the array with pointers from the end
+      --num_goal_handles;
       action_server->impl->goal_handles[i] = action_server->impl->goal_handles[num_goal_handles];
-      action_server->impl->goal_handles[num_goal_handles--] = NULL;
```

Comment:
Is this change desired? You previously accessed, then decremented but now you decrement and the access `num_goal_handles`. I guessing maybe, if you're removing the last element in the list of goal handles, so `num_goal_handles - 1` is both the last index and the new length, whereas before you were probably accessing out of bounds (`[len]` rather than `[len - 1]`)?

Reply:
This one is a fix for an out of bounds error. I think the change is right, but maybe I haven't looked at this closely enough.

In the previous code when there is just one goal handle and it is being expired it would:

```
// num_goal_handles = 1, size of goal_handles = 1
...goal_handles[0] = ...goal_handles[1];  // read out of bounds
...goal_handles[1] = NULL;  // write out of bounds
```

If there were two goal handles and the first is being expired then it would
```
// num_goal_handles = 2, size of goal_handles = 2
// First iteration i = 0
...goal_handles[0] = ...goal_handles[2];  // read out of bounds
...goal_handles[2] = NULL; num_goal_handles -= 1;  // write out of bounds
```


The same two cases in the new code
```
// num_goal_handles = 1, size of goal_handles = 1
num_goal_handles -= 1;
...goal_handles[0] = ...goal_handles[0];  // noop
...goal_handles[0] = NULL;
```

```
// num_goal_handles = 2, size of goal_handles = 2
// First iteration i = 0
num_goal_handles -= 1;
...goal_handles[0] = ...goal_handles[1];
...goal_handles[1] = NULL;
```